From 4728ab1a3302abd91869fe51e0bd7f8d6be2c64b Mon Sep 17 00:00:00 2001
From: Leorize <alaviss@users.noreply.github.com>
Date: Fri, 22 Dec 2017 15:21:32 +0700
Subject: add Haiku support

use find_directory for settings

use int typedefs from SupportDefs instead

use fallback ffsll

diff --git a/lrzip_private.h b/lrzip_private.h
index d524b89..d02216e 100644
--- a/lrzip_private.h
+++ b/lrzip_private.h
@@ -88,6 +88,10 @@ void *alloca (size_t);
 # endif
 #endif
 
+#ifdef __HAIKU__
+#include <SupportDefs.h>
+#endif
+
 #ifndef MD5_DIGEST_SIZE
 # define MD5_DIGEST_SIZE 16
 #endif
@@ -158,7 +162,7 @@ extern int errno;
 #define unlikely(x)	__builtin_expect(!!(x), 0)
 #define __maybe_unused	__attribute__((unused))
 
-#if defined(__MINGW32__) || defined(__CYGWIN__) || defined(__ANDROID__) || defined(__APPLE__) || defined(__OpenBSD__)
+#if defined(__MINGW32__) || defined(__CYGWIN__) || defined(__ANDROID__) || defined(__APPLE__) || defined(__OpenBSD__) || defined(__HAIKU__)
 # define ffsll __builtin_ffsll
 #endif
 
diff --git a/util.c b/util.c
index 60679ae..0d10cc8 100644
--- a/util.c
+++ b/util.c
@@ -60,6 +60,11 @@
 #endif
 #include <inttypes.h>
 
+#ifdef __HAIKU__
+#include <FindDirectory.h>
+#include <fs_info.h>
+#endif
+
 /* Macros for testing parameters */
 #define isparameter( parmstring, value )	(!strcasecmp( parmstring, value ))
 #define iscaseparameter( parmvalue, value )	(!strcmp( parmvalue, value ))
@@ -209,7 +214,13 @@ bool get_rand(rzip_control *control, uchar *buf, int len)
 bool read_config(rzip_control *control)
 {
 	/* check for lrzip.conf in ., $HOME/.lrzip and /etc/lrzip */
+#ifndef __HAIKU__
 	char *HOME, homeconf[255];
+#else
+       dev_t volume;
+       char buffer[B_PATH_NAME_LENGTH];
+       char confpath[B_PATH_NAME_LENGTH + B_FILE_NAME_LENGTH];
+#endif
 	char *parametervalue;
 	char *parameter;
 	char line[255];
@@ -218,6 +229,7 @@ bool read_config(rzip_control *control)
 	fp = fopen("lrzip.conf", "r");
 	if (fp)
 		fprintf(control->msgout, "Using configuration file ./lrzip.conf\n");
+#ifndef __HAIKU__
 	if (fp == NULL) {
 		HOME=getenv("HOME");
 		if (HOME) {
@@ -232,6 +244,25 @@ bool read_config(rzip_control *control)
 		if (fp)
 			fprintf(control->msgout, "Using configuration file /etc/lrzip/lrzip.conf\n");
 	}
+#else
+       if (fp == NULL) {
+               volume = dev_for_path("/boot");
+               if (find_directory(B_USER_SETTINGS_DIRECTORY, volume, false, buffer, sizeof(buffer)) == B_OK) {
+                       snprintf(confpath, sizeof(confpath), "%s/lrzip/lrzip.conf", buffer);
+                       fp = fopen(confpath, "r");
+                       if (fp)
+                               fprintf(control->msgout, "Using configuration file %s\n", confpath);
+               }
+       }
+       if (fp == NULL) {
+               if (find_directory(B_SYSTEM_SETTINGS_DIRECTORY, volume, false, buffer, sizeof(buffer)) == B_OK) {
+                       snprintf(confpath, sizeof(confpath), "%s/lrzip/lrzip.conf", buffer);
+                       fp = fopen(confpath, "r");
+                       if (fp)
+                               fprintf(control->msgout, "Using configuration file %s\n", confpath);
+               }
+       }
+#endif
 	if (fp == NULL)
 		return false;
 
-- 
2.54.0

From b83791d485fed1cb8deacd7d078e5c313887fddc Mon Sep 17 00:00:00 2001
From: Luc Schrijvers <begasus@gmail.com>
Date: Fri, 2 Jan 2026 12:58:55 +0100
Subject: Haiku doesn't have PRIO_MIN(?)


diff --git a/main.c b/main.c
index 7467905..70cf876 100644
--- a/main.c
+++ b/main.c
@@ -492,11 +492,15 @@ int main(int argc, char *argv[])
 		case 'N':
 			nice_set = true;
 			control->nice_val = strtol(optarg, &endptr, 10);
-			if (control->nice_val < PRIO_MIN || control->nice_val > PRIO_MAX)
-				failure("Invalid nice value (must be %d...%d)\n", PRIO_MIN, PRIO_MAX);
-			if (*endptr)
-				failure("Extra characters after nice level: \'%s\'\n", endptr);
-			break;
+			#if defined(__HAIKU__)
+				return false;
+			#else
+				if (control->nice_val < PRIO_MIN || control->nice_val > PRIO_MAX)
+					failure("Invalid nice value (must be %d...%d)\n", PRIO_MIN, PRIO_MAX);
+				if (*endptr)
+					failure("Extra characters after nice level: \'%s\'\n", endptr);
+				break;
+			#endif
 		case 'o':
 			if (control->outdir)
 				failure("Cannot have -o and -O together\n");
-- 
2.54.0


From bb497787bb1a3c197648f311a0728becf725d43b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?J=C3=A9r=C3=B4me=20Duval?= <jerome.duval@gmail.com>
Date: Tue, 22 Sep 2026 16:57:33 +0200
Subject: define unix on Haiku


diff --git a/configure.ac b/configure.ac
index 263a37e..fb07f40 100644
--- a/configure.ac
+++ b/configure.ac
@@ -59,7 +59,7 @@ esac
 # GCC and Linux clang predefine it, but Darwin clang does not, sending macOS
 # builds into the Windows branch (#include <windows.h>).
 case "${host_os}" in
-  darwin*)
+  darwin*|haiku*)
     CPPFLAGS="${CPPFLAGS} -Dunix"
     ;;
 esac
-- 
2.54.0

